agent/agent_endpoint.go

```diff
@@ -166,6 +167,55 @@ func (s *HTTPHandlers) AgentMetrics(resp http.ResponseWriter, req *http.Request)
 	return s.agent.baseDeps.MetricsHandler.DisplayMetrics(resp, req)
 }
 
+func (s *HTTPHandlers) AgentMetricsStream(resp http.ResponseWriter, req *http.Request) (interface{}, error) {
+	// Fetch the ACL token, if any, and enforce agent policy.
+	var token string
+	s.parseToken(req, &token)
+	rule, err := s.agent.delegate.ResolveTokenAndDefaultMeta(token, nil, nil)
+	if err != nil {
+		return nil, err
+	}
+	if rule != nil && rule.AgentRead(s.agent.config.NodeName, nil) != acl.Allow {
```

Comment:
If the `rule` here is actually `nil`, the ACL check is skipped. Is there any case where this could actually happen? Would it make sense to add the `nil` check separately?

Reply:
Excellent question! This `rule != nil` has given me pause as well, and it's something I'd like to change. But I believe it is correct given the way the code works today. I think basically everywhere in the code base that checks ACLs does this very same check (`git grep 'rule != nil'` returns quite a few lines that look exactly like this).

The reason can be seen here: https://github.com/hashicorp/consul/blob/v1.10.0/agent/consul/acl.go#L1187-L1190

When ACLs are disabled we return a nil `acl.Authorizer` (`rule` in this code is an authorizer, I believe the name comes from the old ACL system, and we probably just left the variable names alone when we updated the ACL system).

I think it would be great to change `ResolveToken` to have it return the [AllowAll](https://github.com/hashicorp/consul/blob/v1.10.0/acl/static_authorizer.go#L229-L232) `Authorizer` instead of `nil`. That would remove the need for a `nil` check and I think would make the code less scary to read.

I think that's a change we could make pretty easily in another PR, and see what others think.